python/ray/tests/kuberay/utils.py

```diff
@@ -227,3 +230,102 @@ def get_raycluster(raycluster: str, namespace: str) -> Dict[str, Any]:
         .strip()
     )
     return yaml.safe_load(get_raycluster_output)
+
+
+def _get_service_port(service: str, namespace: str, target_port: int) -> int:
+    """Given a K8s service and a port targetted by the service, returns the
+    corresponding port exposed by the service.
+
+    Args:
+        service: Name of a K8s service.
+        namespace: Namespace to which the service belongs.
+        target_port: Port targeted by the service.
+
+    Returns:
+        service_port: The port exposed by the service.
+    """
+    service_str = (
+        subprocess.check_output(
+            ["kubectl", "-n", namespace, "get", "service", service, "-o", "yaml"]
+        )
+        .decode()
+        .strip()
+    )
+    service_dict = yaml.safe_load(service_str)
+    service_ports: List = service_dict["spec"]["ports"]
+    matching_ports = [
+        port for port in service_ports if port["targetPort"] == target_port
+    ]
+    assert matching_ports
+    service_port = matching_ports[0]["port"]
+    return service_port
+
+
+@contextlib.contextmanager
+def _kubectl_port_forward(
+    service: str, namespace: str, target_port: int, local_port: Optional[int] = None
+) -> Generator[int, None, None]:
+    """Context manager which creates a kubectl port-forward process targeting a
+    K8s service.
+
+    Terminates the port-forwarding process upon exit.
+
+    Args:
+        service: Name of a K8s service.
+        namespace: Namespace to which the service belongs.
+        target_port: The port targeted by the service.
+        local_port: Forward from this port. Optional. By default, uses the port exposed
+        by the service.
+
+    Yields:
+        The local port. The service can then be accessed at 127.0.0.1:<local_port>.
+    """
+    # First, figure out which port the service exposes for the given target port.
+    service_port = _get_service_port(service, namespace, target_port)
+    if not local_port:
+        local_port = service_port
+
+    process = subprocess.Popen(
+        [
+            "kubectl",
+            "-n",
+            namespace,
+            "port-forward",
+            f"service/{service}",
+            f"{local_port}:{service_port}",
+        ]
+    )
+
+    def terminate_process():
+        process.terminate()
```

Comment:
Will this give an error if terminated twice? If yes, we might want to protect against it, so there are no ugly error messages on shutdown :)

Reply:
Hey, great point.
I'm pretty sure `process.terminate()`, but let me confirm and add a comment.

Reply:
Made the cleanup extra carefuly and commented that running it multiple times is unproblematic.